File: 2246-longest-path-with-different-adjacent-characters/2246-longest-path-with-different-adjacent-characters.py
class Solution:
    def longestPath(self, parent: List[int], s: str) -> int:
        child = defaultdict(list)
        n = len(parent)
        for i in range(1, n):
            child[parent[i]].append(i)

        res = [0]
        def dfs(cur):
            a,b = 0,0
            for ch in child[cur]:
                c = dfs(ch)
                if s[ch] != s[cur]:
                    if c > a:
                        b = a
                        a = c
                    elif c > b:
                        b = c
            
            res[0] = max(res[0], a+b+1)
            return a+1
        dfs(0)
        return res[0]
            

File: 2246-longest-path-with-different-adjacent-characters/2246-longest-path-with-different-adjacent-characters.cpp
class Solution {
public:
    int longestPath(vector<int>& parent, string s) {
        int n=s.size();
        vector<int> indegree(n,0);
        for(int i=0;i<n;i++){
            if(parent[i]!=-1){
                indegree[parent[i]]++;
            }
        }
        queue<int> q;
        for(int i=0;i<n;i++){
            if(indegree[i]==0){
                q.push(i);
            }
        }
        vector<int> maxpath(n,1);
        int ans=1;
        while(!q.empty()){
            int u=q.front();
            q.pop();
            
            int v=parent[u];
            if(v==-1){
                continue;
            }
            indegree[v]--;
            if(s[u]!=s[v]){
                ans=max(ans,maxpath[v]+maxpath[u]);
                maxpath[v]=max(maxpath[v],1+maxpath[u]);
            }
            if(indegree[v]==0){
                q.push(v);
            }
        }
        return ans;
    }
};